blog/content/post/2021-12-02-meta-nezha.md

```diff
@@ -0,0 +1,162 @@
+cover: /img/nezha-logo.png
+author: cezary.sobczak
+layout: post
+published: true
+date: 2021-12-02
+archives: "2021"
+
+tags:
+  - risc-v
+  - sbc
+  - linux
+  - u-boot
+  - boot0
+  - opensbi
+  - bootloader
+  - logs
+  - SoC
+  - sdcard
+  - yocto
+  - layer
+  - meta
+categories:
+  - Firmware
+  - IoT
+  - Miscellaneous
+  - OS Dev
+  - App Dev
+  - Security
+  - Manufacturing
+
+---
+
+## Introduction
+
+As it was mentioned in first [post](https://blog.3mdeb.com/2021/2021-11-19-nezha-riscv-sbc-first-impression/)
+I'm interested in support for this board in Yocto Project, and this post will
+show you what I have achieved in this field.
+
+I didn't mention that earlier but the idea of creating the `meta-nezha` yocto
+layer was born in my head when I was thinking about the master thesis subject.
+So the whole project was created for educational purposes to share knowledge
+about Yocto and RISC-V.
+
+Whole code of the meta layer you can find at [github](https://github.com/Cezarus27/meta-nezha). It should be mentioned that `meta-nezha` is really early support so
+the recipe's code isn't that clear and for now, it isn't prepared to be
+upstreamed.
+
+## Nezha D1 meta layer construction
+
+Nezha D1 layer is using the following Yocto Project meta layers in `hardknott`
+version:
+
+* [poky](https://git.yoctoproject.org/git/poky)
+
+* [meta-openembedded](https://git.openembedded.org/meta-openembedded)
+
+* [meta-riscv](https://github.com/riscv/meta-riscv.git)
+
+The structure of a `meta-nezha` is divided into two main parts:
+
+* `meta-nezha-bsp` - this layer contains recipes for `boot0`, `u-boot`,
+  `OpenSBI` and `linux kernel`. It also contains machine configuration which
+  set/unset or enable/disable the key features of the board,
+
+* `meta-nezha-distro` - contain a recipe for the minimal image, configuration of the
+  system eg. what init manager is used and `wks` file which is used to create
+  image file which can be flashed to the SD card.
+
+Key repositories used by `meta-nezha-bsp` recipes are forks of repositories
+patched / created by **smaeul**:
+
+* [OpenSBI](https://github.com/Cezarus27/opensbi/tree/d1-wip): `d1-wip` branch
+
+* [u-boot](https://github.com/Cezarus27/u-boot/tree/d1-wip): `d1-wip` branch
+
+* [Linux](https://github.com/Cezarus27/linux/tree/riscv/d1-wip): `riscv/d1-wip`
+  branch
+
+* [boot0](https://github.com/Cezarus27/sun20i_d1_spl): `mainline` branch
+
+> Note: I decide to fork the `smaeul` repositories because the original repos
+> are often updated and rebased to the newest branches from upstream
+> repositories and it often caused problems with the proper build of the yocto
+> image.
+
+Main work to do was to adopt the bootflow of Nezha to the Yocto Project. In this
+process, the `boot0` recipe was created from the ground. That wasn't so smooth
+to integrate `boot0` compilation flow with the Yocto build engine due to missing
+some headers. So the patch for `boot0` Makefile was provided also and it is
```

Comment:
It is known issue on upstream? Maybe we could contribute?

Reply:
I think it is specific for building in yocto because the location of needed files in `workdir` is different from manual compilation.